code/ai/aigoals.cpp

```diff
@@ -134,28 +136,41 @@ const char *Ai_goal_text(int goal)
 	return NULL;
 }
 
-// function to maybe add the form on my wing goal for a player's starting wing.  Called when a player wing arrives.
-void ai_maybe_add_form_goal( wing *wingp )
+void ai_maybe_add_form_goal(wing* wingp)
 {
 	int j;
 
 	// iterate through the ship_index list of this wing and check for orders.  We will do
 	// this for all ships in the wing instead of on a wing only basis in case some ships
 	// in the wing actually have different orders than others
-	for ( j = 0; j < wingp->current_count; j++ ) {
-		ai_info *aip;
+	for (j = 0; j < wingp->current_count; j++) {
+		ai_info* aip;
 
-		Assert( wingp->ship_index[j] != -1 );						// get Allender
+		Assert(wingp->ship_index[j] != -1);						// get Allender
 
 		aip = &Ai_info[Ships[wingp->ship_index[j]].ai_index];
 		// don't process Player_ship
-		if ( aip == Player_ai )
+		if (aip == Player_ai || (MULTIPLAYER_MASTER && Objects[Ships[aip->shipnum].objnum].flags[Object::Object_Flags::Player_ship])) {
```

Comment:
Does the `MULTIPLAYER_MASTER` check need to be here?  A player ship shouldn't be processed regardless of what type of build is running.

Reply:
The check for multiplayer guarantees two things: one, that the game is in multiplayer so SP will not be affected, and two, that only the server will make these changes.  Since the server takes care of the all of the ai (it gets sent via object update packets), doing otherwise could cause server and client to go out of sync, albeit probably not noticeably.

Reply:
If only the server should make these changes, then shouldn't you instead put
``
if (!MULTIPLAYER_MASTER)
    return;
```
at the very top of the function?

Reply:
Hmmmmm.... Yes, that makes sense.  Perhaps in the future someone can go through and further optimize the ai code for clients for similar functions.